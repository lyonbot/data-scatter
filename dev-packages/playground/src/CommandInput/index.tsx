import * as React from 'react';
import CodeMirror from 'codemirror';
import 'codemirror/lib/codemirror.css'
import 'codemirror/addon/display/placeholder'
import 'codemirror/addon/hint/show-hint'
import 'codemirror/addon/hint/show-hint.css'
import 'codemirror/addon/hint/javascript-hint'
import 'codemirror/mode/javascript/javascript'
import './style.scss'

type Props = React.PropsWithoutRef<{
  value?: string
  placeholder?: string
  onReady?: (cm: CodeMirror.Editor) => void
  onChange?: (code: string) => void
  onKeyDown?: (cm: CodeMirror.Editor, event: KeyboardEvent) => void
  onKeyUp?: (cm: CodeMirror.Editor, event: KeyboardEvent) => void
  onSubmit?: (code: string, fn: () => Promise<any>) => Promise<void> | void;
  /** separated by comma */
  functionArgumentList?: string | string[]
}>;

export const CommandInput = React.memo((_props: Props) => {
  const $cm = React.useRef<CodeMirror.Editor | null>(null)
  const $el = React.useRef<HTMLDivElement>(null)
  const props = React.useRef<Props>(_props)
  const [, setReady] = React.useState(false)

  React.useEffect(() => {
    const cm = CodeMirror($el.current!, {
      // @ts-ignore
      mode: { name: 'javascript', globalVars: true },
      hintOptions: {
        globalScope: window
      } as any,
      lineNumbers: false,
      lineWrapping: true,
      viewportMargin: Infinity,
    })
    $cm.current = cm
    setReady(true)

    cm.addKeyMap({
      'Ctrl-Enter': 'newlineAndIndent',
      'Cmd-Enter': 'newlineAndIndent',
      "Ctrl-Space": "autocomplete",
      'Enter': () => {
        const value = cm.getValue();
        if (!value.trim()) return;

        if (cm.getOption('readOnly')) return;
        cm.setOption('readOnly', true);

        (async () => {
          // eslint-disable-next-line @typescript-eslint/ban-types
          let fn: Function | null = null;

          try {

            // make an async function

            try {
              fn = new Function(`return (async function(${props.current.functionArgumentList || ''}){ return (${value}\n); }).apply(this, arguments)`)
            } catch {
              fn = new Function(`return (async function(${props.current.functionArgumentList || ''}){ ${value}\n; }).apply(this, arguments)`)
            }

            // callback

            await props.current.onSubmit?.(value, fn as () => Promise<any>)
            cm.setValue('')
          } catch (err) {
            console.error('Failed to execute', err)
            cm.execCommand('newlineAndIndent')
          }
          cm.setOption('readOnly', false)
        })()
      }
    })

    let isHintShown = false
    cm.on('startCompletion', () => void (isHintShown = true));
    cm.on('endCompletion', () => void (isHintShown = false));

    cm.on('keydown', (instance, event) => {
      if (!isHintShown) {
        props.current.onKeyDown?.(cm, event)
      }
    })

    cm.on('keyup', (instance, event) => {
      if (!isHintShown) {
        props.current.onKeyUp?.(cm, event)
        if (event.code === 'Period') cm.execCommand('autocomplete')
      }
    })

    cm.on('changes', () => {
      props.current.onChange?.(cm.getValue())
    })

    props.current.onReady?.(cm)
  }, [])

  const cm = $cm.current
  props.current = _props

  React.useEffect(() => { cm?.setValue(_props.value || '') }, [cm, _props.value])
  React.useEffect(() => { cm?.setOption('placeholder', _props.placeholder || '') }, [cm, _props.placeholder])

  return <div className="commandInput" ref={$el}></div>
})