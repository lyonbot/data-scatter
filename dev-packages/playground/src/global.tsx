import { createSchemaRegistry, FromSchemaRegistry, ScatterStorage } from 'data-scatter';
import { loadIntoStorage, dumpOneNode, dumpNodesFromStorage, NodeContentObserver, Watcher, NodeInfo } from 'data-scatter'

const registry = createSchemaRegistry({
  // ----------------------------------------------------------------
  // define a schema in JSON-Schema like format
  person: {
    type: 'object',
    properties: {
      // you can define type in JSON Schema
      name: { type: 'string' },

      // or just refer to a defined type
      father: 'person',
      mother: 'person',
      employer: 'entrepreneur',

      // you can also do it in a nested type
      // ( see `items` )
      children: { type: 'array', items: 'person' },
    },
  },

  // ----------------------------------------------------------------
  // extend an existing object schema
  entrepreneur: {
    type: 'object',
    title: 'A person with goals!',
    extends: ['person'],
    properties: {
      goals: { type: 'array', items: { type: 'string' } },
      employer: null, // not inherit this property
    },
  },

  task: {
    type: 'object',
    properties: {
      name: { type: 'string' },
      executor: { type: 'string' },
      subTasks: { type: 'array', items: 'task' },
      assignee: 'person'
    }
  },
});

export const storage = new ScatterStorage({ schemaRegistry: registry });

const sampleData: FromSchemaRegistry<typeof registry, 'task'> = {
  name: 'go shopping',
  assignee: { name: 'Tony' },
  subTasks: [
    { name: 'buy flowers' },
    { name: 'buy chocolate' },
  ]
};

const doc = storage.create('task', sampleData);

export const global = {
  registry,
  storage,
  doc,
  createSchemaRegistry,
  ScatterStorage,
  loadIntoStorage,
  dumpOneNode,
  dumpNodesFromStorage,
  NodeContentObserver,
  Watcher,
};
