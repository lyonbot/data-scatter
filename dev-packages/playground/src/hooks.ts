import { useMemo, useReducer, useRef } from "react";

/**
 * Get a proxy to a chaging object / function and the proxy's ref is always the same one.
 * 
 * @param updateWhenSet do a forceUpdate when assigning value to the proxy. default is false
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export function useLast<T extends Function | Object>(value: T, updateWhenSet?: boolean): T {
  const lastRef = useRef(value);
  const updateFlag = useRef(false)
  const forceUpdate = useForceUpdate()

  const h = useMemo(() => {
    if (typeof value === 'function') {
      return function (this: any, ...args: any[]) { return (lastRef.current as any).apply(this, args); }
    }

    const mayCall = () => updateFlag.current && forceUpdate()

    return new Proxy(Object.create(Object.getPrototypeOf(value)), {
      get(_, p) { return Reflect.get(lastRef.current, p) },
      set(_, p, v) { mayCall(); return Reflect.set(lastRef.current, p, v) },
      deleteProperty(_, p) { mayCall(); return Reflect.deleteProperty(lastRef.current, p) },
      ownKeys() { return Reflect.ownKeys(lastRef.current) },
      has(p) { return Reflect.has(lastRef.current, p) },
    })
  }, [typeof value, Array.isArray(value)])

  lastRef.current = value
  updateFlag.current = !!updateWhenSet

  return h as T
}

export function useForceUpdate() {
  const [, r] = useReducer((x) => (x + 1) % 0xffffff, 0)
  return r as () => void
}
