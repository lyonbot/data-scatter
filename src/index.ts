export * from './schema'
export * from './scatter'
export * from './types'
export { EventEmitter, combineFunctions } from './EventEmitter'
