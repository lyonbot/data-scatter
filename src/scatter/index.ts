export { ScatterStorage } from './storage'
export { loadIntoStorage, dumpOneNode, dumpNodesFromStorage } from './loadDump'
export { arrayify, makeEmptyLike } from './utils'
export { NodeContentObserver, Watcher } from './observe'

export type { DumpedNodeData, LoadIntoStorageOptions, LoadIntoStorageResponse } from './loadDump'
export type { NodeInfo } from './NodeInfo'
export type { AutoScatterEvents, ScatterStorageInitOptions, NodeIdGenerator, NodeWriteAccessAction } from './storage'
export type { OneOrMany } from './utils'
export type { WalkStepInfo, WalkOptions, WalkCallbackResponse } from './walk'
export type { NodeContentObserverEvents } from './observe'
