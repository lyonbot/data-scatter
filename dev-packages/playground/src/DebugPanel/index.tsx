import * as React from 'react';
import { toPath, get } from 'lodash';
import classnames from 'classnames';
import { CommandInput } from '../CommandInput';
import { MyInspector } from '../MyInspector';
import { CodeSnippet } from './runmode';
import "./style.scss"

// check if a string is path
const rePathValidator = /^\s*[a-z_$][\w$]*(\.[a-z_$][\w$]*|\[(-?\d+|'[^']*'|"[^"]*")\])*\s*$/i

type Props = React.HTMLProps<HTMLDivElement> & {
  /** a notice text */
  notice?: React.ReactNode
  placeholder?: string
}

export const DebugPanel = (_props: Props) => {
  const logsDiv = React.useRef<HTMLDivElement>(null)
  const { notice = 'ðŸ‘† Try executing some JavaScript above', placeholder, ...otherProps } = _props

  //----------------------------------------------------------------

  const recentCodes = React.useMemo(() => ({
    list: [] as string[],     // in reversed order: latest first
    index: 0
  }), [])
  const [records, updateRecords] = React.useState<React.ReactElement[]>([]);
  const clearRecords = React.useCallback(() => {
    recentCodes.index = 0
    recentCodes.list.length = 0
    updateRecords([]);
  }, []);

  const handleCommandSubmit = React.useCallback(async (code: string, fn: () => Promise<any>) => {
    const log = (...items: React.ReactElement[]) => updateRecords(x => {
      const origLength = x.length;
      return x.concat(items.map((item, index) => React.cloneElement(item, {
        key: index + origLength,
        className: 'debugPanel-row ' + (item.props.className || '')
      })));
    });

    log(<div>
      <div className="debugPanel-indicator isGrey">&raquo;</div>
      <pre className='debugPanel-pre'><CodeSnippet code={code} /></pre>
    </div>);
    try {
      console.log("%c%s", "color:#35f", code)
      recentCodes.list.unshift(code)
      const result = await fn();
      if (typeof result !== 'undefined') {
        console.log(result)
        log(<div><MyInspector data={result} /></div>);
      }
    } catch (error) {
      console.error(error)
      log(<div className="isError">
        <div className="debugPanel-indicator">
          <span className="debugPanel-errorMark">!</span>
        </div>
        <MyInspector data={error} />
      </div>);
    }
  }, []);

  React.useEffect(() => {
    const div = logsDiv.current
    if (!div) return;

    div.scrollTo(0, div.scrollHeight)
  }, [records.length])

  //----------------------------------------------------------------

  const [code, setCode] = React.useState('')
  const isCodePath = React.useMemo(() => (rePathValidator.test(code) && toPath(code.trim())), [code])
  const instantChildren = React.useMemo(() => {
    if (!isCodePath) return null

    const value = get(window, isCodePath)
    return <MyInspector data={value} />
  }, [isCodePath]) || (code ? "Press Enter to Execute" : notice)

  //----------------------------------------------------------------

  return <div {...otherProps} className={classnames("debugPanel", otherProps.className)}>
    <div className="debugPanel-logs" ref={logsDiv}>
      {records}
    </div>
    <div className="debugPanel-input">
      <div className="debugPanel-indicator isBlue">&raquo;</div>
      <CommandInput
        placeholder={placeholder}
        onKeyDown={(cm, event) => {
          if (!cm.somethingSelected()) {
            let delta = 0;
            if (event.code === 'ArrowUp' && cm.getCursor().line === 0) delta = 1;
            if (event.code === 'ArrowDown' && cm.getCursor().line === cm.lineCount() - 1) delta = -1;

            if (delta) {
              const code = recentCodes.list[recentCodes.index]
              cm.setValue(code)
              document.addEventListener('keyup', () => cm.execCommand('selectAll'), { once: true })
              recentCodes.index = (recentCodes.index + delta + recentCodes.list.length) % recentCodes.list.length
            } else {
              recentCodes.index = 0
            }
          }
        }}
        onSubmit={handleCommandSubmit}
        onChange={setCode}
      />
    </div>
    <div className="debugPanel-instant">
      {instantChildren}
    </div>
  </div>;
};
