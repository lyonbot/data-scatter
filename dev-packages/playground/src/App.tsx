import * as React from 'react';
import ReactMarkdown from 'react-markdown';
import { DebugPanel, DebugPanelRef } from './DebugPanel';
import { NodeInfoView } from './NodeInfoView';
import introMarkdown from './intro.md?raw';
import { storage } from "./global";

import 'github-markdown-css/github-markdown-light.css'
import './index.css';
import { hoveringItemAtom, useAtom } from './state';
import { MyInspector } from './MyInspector';

// ----------------------------------------------------------------
const defaultOpts = {
  showRefNodeId: false
};

const SideBar = () => {
  const [active] = useAtom(hoveringItemAtom)

  return <div>
    <MyInspector data={active.data} />
  </div>
}

export const App = () => {
  const debugPanelRef = React.useRef<DebugPanelRef>(null);
  const [nodes, setNodes] = React.useState(() => Array.from(storage.nodes.values()));
  const [opts, updateOpts] = React.useState(defaultOpts);

  React.useEffect(() => {
    storage.on('nodeCreated', nodeInfo => {
      setNodes(arr => arr.concat(nodeInfo));
    });

    const addRecord = debugPanelRef.current!.addRecord;
    addRecord(<div className='py-8'>
      <ReactMarkdown children={introMarkdown} className="markdown-body" />
    </div>);
  }, []);

  return <div className="flex absolute inset-0 gap-4">
    <div className="flex-1 grow-[2] max-h-full">
      <DebugPanel placeholder="data.xxx" className='max-h-full' ref={debugPanelRef} />
    </div>

    <div className="flex-1 flex flex-col">
      <div className="p-4">
        test

        <SideBar />

        <label>
          <input type="checkbox" checked={opts.showRefNodeId} onChange={() => updateOpts(x => ({ ...x, showRefNodeId: !x.showRefNodeId }))} />
          showRefNodeId
        </label>
      </div>
      <div className="flex-1 overflow-auto">
        <div>
          {nodes.map(nodeInfo => <NodeInfoView
            key={nodeInfo.id}
            nodeInfo={nodeInfo}
            showRefNodeId={opts.showRefNodeId} />)}
        </div>
      </div>
    </div>
  </div>;
};
