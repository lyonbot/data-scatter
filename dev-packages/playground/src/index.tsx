import * as React from 'react';
import * as ReactDOM from 'react-dom/client';
import { createSchemaRegistry, FromSchemaRegistry, ScatterStorage, loadIntoStorage, dumpOneNode, dumpNodesFromStorage, NodeContentObserver, Watcher } from 'data-scatter'
import { DebugPanel } from './DebugPanel';
import { MyInspector } from './MyInspector';

declare module "data-scatter" {

  // add basic fields for schemas

  export interface SchemaBase {
    description?: string
    tooltip?: string
  }

  // add basic types
  // so you can use them like { type: "integer" }

  export interface PrimitiveTypeLUT {
    integer: number,  // { type: "integer" } -> javascript number
  }
}

const registry = createSchemaRegistry({
  // ----------------------------------------------------------------
  // define a schema in JSON-Schema like format
  person: {
    type: 'object',
    properties: {
      // you can define type in JSON Schema
      name: { type: 'string' },

      // or just refer to a defined type
      father: 'person',
      mother: 'person',
      employer: 'entrepreneur',

      // you can also do it in a nested type
      // ( see `items` )
      children: { type: 'array', items: 'person' },
    },
  },

  // ----------------------------------------------------------------
  // extend an existing object schema
  entrepreneur: {
    type: 'object',
    title: 'A person with goals!', // <- extra notes
    extends: ['person'], // <- inherit properties from `person`
    properties: {
      goals: { type: 'array', items: { type: 'string' } }, // add string[]
      employer: null, // not inherit this property
    },
  },

  task: {
    type: 'object',
    properties: {
      name: { type: 'string' },
      executor: { type: 'string' },
      subTasks: { type: 'array', items: 'task' },
      assignee: 'person'
    }
  },
});

const storage = new ScatterStorage({ schemaRegistry: registry })

const sampleData: FromSchemaRegistry<typeof registry, 'task'> = {
  name: 'go shopping',
  assignee: { name: 'Tony' },
  subTasks: [
    { name: 'buy flowers' },
    { name: 'buy chocolate' },
  ]
}

const doc = storage.create('task', sampleData)

const global = {
  registry,
  storage,
  doc,
  createSchemaRegistry,
  ScatterStorage,
  loadIntoStorage,
  dumpOneNode,
  dumpNodesFromStorage,
  NodeContentObserver,
  Watcher,
}
Object.assign(window, global)


const root = ReactDOM.createRoot(document.getElementById('app')!);
root.render(<div>

  <DebugPanel placeholder="data.xxx" />
  <MyInspector data={global} onClick={(x) => console.log(x)} />

</div>)
