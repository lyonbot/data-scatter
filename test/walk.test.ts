import { loadIntoStorage, ScatterStorage, WalkCallbackResponse, WalkOptions, WalkStepInfo } from '../src/scatter'
import { getSchemaRegistry, sampleDumpedData1 } from './fixture'

describe('walk', () => {
  const visitedLogs: any[] = [];
  const writeLog = (step: WalkStepInfo) => visitedLogs.push(`id=${step.nodeId}  path=${JSON.stringify(step.path)}  isVisited=${step.isVisited}`)

  beforeEach(() => {
    visitedLogs.splice(0)
  })

  test.each([
    {},
    { mode: 'DFS' as const },
    { async: true }
  ])('basic walk, %p', async (testCaseOpts) => {
    const storage = new ScatterStorage({ schemaRegistry: getSchemaRegistry() });
    loadIntoStorage({
      storage,
      nodes: sampleDumpedData1
    })

    const task1VisitedLog: WalkStepInfo[] = []

    const visitor = (step: WalkStepInfo): WalkCallbackResponse => {
      if (step.nodeId === 'task1') task1VisitedLog.push(step)
      writeLog(step)

      if (step.isVisited) return 'skip-children';
    }

    const walkOpt: WalkOptions = {
      startPath: ['myTask'],
      mode: testCaseOpts.mode,
    }

    if (testCaseOpts.async) {
      const resp = storage.walk('task1', async o => visitor(o), walkOpt);
      expect(resp).toBeInstanceOf(Promise);
      await resp
    } else {
      storage.walk('task1', visitor, walkOpt);
    }

    // ----------------------------
    expect(visitedLogs).toMatchSnapshot('visitedLogs')
  })

  test('callback returns', () => {
    const storage = new ScatterStorage({ schemaRegistry: getSchemaRegistry() });
    loadIntoStorage({
      storage,
      nodes: sampleDumpedData1
    })

    const visitor = jest.fn((step: WalkStepInfo): WalkCallbackResponse => {
      writeLog(step)

      if (step.nodeId === 'task1') return { only: ['meta'] }
      if (step.nodeId === 'anonymousObject1') return { skips: ['task'] }
    })

    storage.walk(['task1', ''], visitor)

    expect(visitedLogs).toMatchInlineSnapshot(`
[
  "id=task1  path=[]  isVisited=0",
  "id=anonymousObject1  path=["meta"]  isVisited=0",
]
`)
  })

  test.each([true, false])('throw error, isAsync = %p', async (isAsync) => {
    const storage = new ScatterStorage({ schemaRegistry: getSchemaRegistry() });
    loadIntoStorage({
      storage,
      nodes: sampleDumpedData1
    })

    const fn = (step: WalkStepInfo) => {
      writeLog(step)

      if (step.nodeId === 'anonymousObject1') throw new Error('bad')
    }

    if (isAsync) {
      const resp = storage.walk('task1', async (x) => fn(x));
      expect(resp).toBeInstanceOf(Promise)
      await expect(resp).rejects.toThrow('bad')
    } else {
      expect(() => storage.walk('task1', fn)).toThrow('bad')
    }

    expect(visitedLogs).toEqual([
      'id=task1  path=[]  isVisited=0',
      'id=array1  path=["subTasks"]  isVisited=0',
      'id=anonymousObject1  path=["meta"]  isVisited=0',
    ])
  })
})
