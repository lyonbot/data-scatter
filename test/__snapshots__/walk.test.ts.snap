// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`walk basic walk, {"async": true}: visitedLogs 1`] = `
[
  "id=task1  path=["myTask"]  isVisited=0",
  "id=array1  path=["myTask","subTasks"]  isVisited=0",
  "id=anonymousObject1  path=["myTask","meta"]  isVisited=0",
  "id=task1  path=["myTask","subTasks",0]  isVisited=1",
  "id=task1  path=["myTask","meta","task"]  isVisited=2",
]
`;

exports[`walk basic walk, {"mode": "DFS"}: visitedLogs 1`] = `
[
  "id=task1  path=["myTask"]  isVisited=0",
  "id=array1  path=["myTask","subTasks"]  isVisited=0",
  "id=task1  path=["myTask","subTasks",0]  isVisited=1",
  "id=anonymousObject1  path=["myTask","meta"]  isVisited=0",
  "id=task1  path=["myTask","meta","task"]  isVisited=2",
]
`;

exports[`walk basic walk, {}: visitedLogs 1`] = `
[
  "id=task1  path=["myTask"]  isVisited=0",
  "id=array1  path=["myTask","subTasks"]  isVisited=0",
  "id=anonymousObject1  path=["myTask","meta"]  isVisited=0",
  "id=task1  path=["myTask","subTasks",0]  isVisited=1",
  "id=task1  path=["myTask","meta","task"]  isVisited=2",
]
`;
